### PR TITLE
Add automation for BZ 2175005, BZ 1955861/1784012

### PR DESCRIPTION
**Pull request by Gauravtalreja1**
QETestCoverage for [BZ 2168967](https://bugzilla.redhat.com/show_bug.cgi?id=2168967) / [BZ 2175005](https://bugzilla.redhat.com/show_bug.cgi?id=2175005) and [BZ 1955861](https://bugzilla.redhat.com/show_bug.cgi?id=1955861) / [BZ 1784012](https://bugzilla.redhat.com/show_bug.cgi?id=1784012)

Dependant PR - https://github.com/SatelliteQE/nailgun/pull/966

### CODE REVIEW

**Comment by Gauravtalreja1** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/{test_provisioningtemplate.py,test_provisioning.py} 
provisioning: true

**Comment by Gauravtalreja1** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/{test_provisioningtemplate.py,test_provisioning.py}
provisioning: true

**Comment by Gauravtalreja1** [comment]
Provisioning passed in PRT 3129 for el7, el8 and only el9 failed due to known issue and is unrelated, so I'd run only ProvisioningTemplates tests to make sure PRT passes.

**Comment by Gauravtalreja1** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_provisioningtemplate.py
nailgun: 966

**Comment by shweta83** [comment]
Description also includes  BZ#2175005 but I don't see it in test case. Can you please add it in description or add in test docstring.

File: tests/foreman/api/test_provisioning.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def test_rhel_pxe_provisioning(
         2. Satellite is able to run REX job on the host
         3. Host is registered to Satellite and subscription status is 'Success'
 
-    :BZ: 2105441
+    :BZ: 2105441, 1955861, 1784012
 
     :customerscenario: true
```
Comment (shweta83) [inline tests/foreman/api/test_provisioning.py:57]: :parametrized: yes

**Comment by Gauravtalreja1** [comment]
> Description also includes BZ#2175005 but I don't see it in test case. Can you please from remove it in description or add in test docstring.

@shweta83 mentioned bz coverage is added in test_provisiining_template.py, and not in provisioning

File: tests/foreman/api/test_provisioning.py (hunk 1 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioning.py:57]: yes

File: tests/foreman/api/test_provisioning.py [hunk 2]
```diff
@@ -121,6 +121,10 @@ def test_rhel_pxe_provisioning(
         provisioning_host.os_version == expected_rhel_version
     ), 'Different than the expected OS version was installed'
 
+    # Verify provisioning log exists on host at correct path
+    assert provisioning_host.execute('test -f /root/install.post.log').status == 0
```
Comment (lhellebr) [inline tests/foreman/api/test_provisioning.py:125]: Should we check for some content in this file? What if it's created but stays empty?

File: tests/foreman/api/test_provisioningtemplate.py [hunk 3]
```diff
@@ -244,16 +244,14 @@ def test_positive_build_pxe_default(self, tftpboot, module_target_sat):
...
+                rendered = module_target_sat.execute(f'cat {template["path"]}').stdout
 
             if template['kind'] == 'iPXE':
                 assert f'{module_target_sat.hostname}/unattended/iPXE' in r.text
             else:
                 assert (
-                    rendered == f'{settings.server.scheme}://'
-                    f'{module_target_sat.hostname} {template["kind"]}'
+                    f'{settings.server.scheme}://{module_target_sat.hostname} {template["kind"]}'
+                    in rendered
```
Comment (lhellebr) [inline tests/foreman/api/test_provisioningtemplate.py:254]: I'm not against this but how is it related the rest of this PR?

File: tests/foreman/api/test_provisioningtemplate.py [hunk 4]
```diff
@@ -406,7 +404,65 @@ def test_positive_template_check_vlan_parameter(
...
+        :customerscenario: true
+        """
+        subnet = module_target_sat.api.Subnet(
+            name=gen_string('alpha'),
+            organization=[module_sca_manifest_org],
+            location=[module_location],
+            network='192.168.0.1',
+            mask='255.255.255.240',
+            boot_mode=boot_mode,
+        ).create()
```
Comment (lhellebr) [inline tests/foreman/api/test_provisioningtemplate.py:444]: Shouldn't this be a fixture?

File: tests/foreman/api/test_provisioningtemplate.py [hunk 5]
```diff
@@ -406,7 +404,65 @@ def test_positive_template_check_vlan_parameter(
...
+        # Verify provision templates for boot_mode in subnet, and check provision logs exists
+        rendered = host.read_template(data={'template_kind': 'provision'})['template']
+        assert f'--bootproto {boot_mode.lower()}' in rendered
+        assert '/root/install.post.log' in rendered
+        assert '/mnt/sysimage/root/install.post.log' not in rendered
+
+        # Verify PXE templates for boot_mode in subnet
+        pxe_templates = ['PXEGrub', 'PXEGrub2', 'PXELinux', 'iPXE']
+        provision_url = f'http://{module_target_sat.hostname}/unattended/provision'
```
Comment (lhellebr) [inline tests/foreman/api/test_provisioningtemplate.py:464]: I don't see BZ 2168967 (escaping of `&`) actually tested here, can you explain?

File: tests/foreman/api/test_provisioning.py (hunk 2 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioning.py:125]: sure, good point, `s/-f/-s` should do that

File: tests/foreman/api/test_provisioningtemplate.py (hunk 3 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioningtemplate.py:254]: yes, its unrelated, but in my last refactor PR I missed it, so updating it rn, looks non-blocking

File: tests/foreman/api/test_provisioningtemplate.py (hunk 4 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioningtemplate.py:444]: it is surely part of setup for the test, but I need this in test as I'm parametrizing it rn. Also if we need this in other tests so will surely consider moving it to global fixtures

File: tests/foreman/api/test_provisioningtemplate.py (hunk 5 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioningtemplate.py:464]: Yes sure, before this BZ, PXEGrub2 template wasn't rendering for UEFI when boot_mode==Static and provisioning fails, but it works now so we're just testing the template render here, and I included all other PXE templates to have better coverage around this scenario.

File: tests/foreman/api/test_provisioning.py (hunk 2 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioning.py:125]: @rplevka Good idea, will open up a separate PR to make it available in artifacts

File: tests/foreman/api/test_provisioningtemplate.py [hunk 6]
```diff
@@ -406,7 +404,65 @@ def test_positive_template_check_vlan_parameter(
...
+        ks_param = provision_url + '?static=1' if boot_mode == 'Static' else provision_url
```
[the remark this answers is not in the document]
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioningtemplate.py:465]: Yes, it'll look cleaner but as there's only single boot_mode param atm, so would defer on doing this but can consider in future , thank you!

File: tests/foreman/api/test_provisioning.py (hunk 1 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioning.py:57]: updated

File: tests/foreman/api/test_provisioningtemplate.py (hunk 4 above)
Comment (lhellebr) [reply tests/foreman/api/test_provisioningtemplate.py:452]: Can't that be resolved by indirect parametrization?

File: tests/foreman/api/test_provisioningtemplate.py (hunk 5 above)
Comment (lhellebr) [reply tests/foreman/api/test_provisioningtemplate.py:472]: In the downloaded template, should the `&` be escaped? Or is it in some other place? If it should be escaped, can you test it, as an addition to testing that the template just renders?

File: tests/foreman/api/test_provisioningtemplate.py (hunk 3 above)
Comment (lhellebr) [reply tests/foreman/api/test_provisioningtemplate.py:254]: I'd prefer unrelated changes to be a separate PR **or** a separate commit, can you do that?

File: tests/foreman/api/test_provisioningtemplate.py (hunk 4 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioningtemplate.py:452]: I just need dummy subnet, so I don't think any test might require that, so I don't think it'll make sense to move it to fixture

**Comment by adarshdubey-star** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_provisioningtemplate.py
nailgun: 966

**Review by adarshdubey-star — APPROVED** [review]
Ack , looks good to me pending comments from @lhelleb & @rplevka

File: tests/foreman/api/test_provisioningtemplate.py (hunk 5 above)
Comment (Gauravtalreja1) [reply tests/foreman/api/test_provisioningtemplate.py:472]: ah, I got you know, so you're mentioning about `options.push("ks=#{foreman_url('provision', static6: '1').gsub("&", "\\\\&")}")` from original PR, so this is adding `'?static=1'` in the end of provision_url, and we're testing the very same URL is rendered correctly in the test at last `assert f'ks={ks_param}' in rendered`

@lhellebr If you still think if I misunderstood this, so please feel free to open a separate PR to add this additional assertion, I'd very much appreciate it, Thank you!

**Review by omkarkhatavkar — APPROVED** [review]

**Pull request merged**